### PR TITLE
Spark Runner: Change to use partitioner in GroupNonMergingWindowsFunctions#groupByKeyInGlobalWindow

### LINKED ISSUE #32608: [Task]: Spark Runner GroupNonMergingWindowsFunctions#groupByKeyInGlobalWindow does not using partitioner
### What needs to happen?

In Spark Runner [`GroupNonMergingWindowsFunctions#groupByKeyInGlobalWindows`](https://github.com/apache/beam/blob/301286fd6a80bfae6f63d4998d002801716eceef/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java#L266) does not use the partitioner passed as a parameter.

### Issue Priority

Priority: 3 (nice-to-have improvement)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Infrastructure
- [X] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by twosom**
**Please** add a meaningful description for your change here

fixes #32608 

This PR contains these changes

- change `GroupNonMergingWindowsFunctions#groupByKeyInGlobalWindow` to use partitioner
- add related test

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by twosom** [comment]
Run Java PreCommit

**Comment by twosom** [comment]
I think task fails is not related to this PR.

**Comment by twosom** [comment]
Run Java PreCommit

**Comment by twosom** [comment]
Run Java PreCommit

File: .github/trigger_files/beam_PostCommit_Java_ValidatesRunner_Spark.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "comment": "Modify this file in a trivial way to cause this test suite to run",
   "https://github.com/apache/beam/pull/31156": "noting that PR #31156 should run this test",
-  "https://github.com/apache/beam/pull/31798": "noting that PR #31798 should run this test"
+  "https://github.com/apache/beam/pull/31798": "noting that PR #31798 should run this test",
+  "https://github.com/apache/beam/pull/32610": "noting that PR #32610 should run this test"
```
Comment (je-ik) [inline .github/trigger_files/beam_PostCommit_Java_ValidatesRunner_Spark.json:5]: I think we can remove these lines, the purpose of the file is to be just touched somehow for the tests to run. We might replace the old comment with a single one.

**Review by je-ik — COMMENTED** [review]
LGTM overall, just one question from higher-level perspective. Given that the only partitioner that is ever passed to this method is returned by
```java
  private static @Nullable Partitioner getPartitioner(SparkTranslationContext context) {
    Long bundleSize =
        context.serializablePipelineOptions.get().as(SparkPipelineOptions.class).getBundleSize();
    return (bundleSize > 0)
        ? null
        : new HashPartitioner(context.getSparkContext().defaultParallelism());
  }
```

does this change have any practical impact? I would expect that the HashPartitioner with default parallelism is what would be used when the partitioner is unused, is this asuumption wrong?

**Comment by twosom** [comment]
> LGTM overall, just one question from higher-level perspective. Given that the only partitioner that is ever passed to this method is returned by
> 
> ```java
>   private static @Nullable Partitioner getPartitioner(SparkTranslationContext context) {
>     Long bundleSize =
>         context.serializablePipelineOptions.get().as(SparkPipelineOptions.class).getBundleSize();
>     return (bundleSize > 0)
>         ? null
>         : new HashPartitioner(context.getSparkContext().defaultParallelism());
>   }
> ```
> 
> does this change have any practical impact? I would expect that the HashPartitioner with default parallelism is what would be used when the partitioner is unused, is this assumption wrong?


Yes, as you mentioned, following the code ultimately leads to using the HashPartitioner.
However, I assumed that the partitioner would be passed as a parameter to the GroupNonMergingWindowsFunctions#groupByKeyInGlobalWindow method, so that it could be used if necessary.
While the current changes may not have a significant practical impact, this decision was made with future extensibility in mind, allowing the use of custom partitioners for specific Apache Beam PTransforms.
Of course, this may not be necessary in the end, but I wanted to leave room for this kind of flexibility.
Thank you.

**Review by kennknowles — APPROVED** [review]
Seems OK to me. I defer to Jan as well.

File: .github/trigger_files/beam_PostCommit_Java_ValidatesRunner_Spark.json (hunk 1 above)
Comment (twosom) [reply .github/trigger_files/beam_PostCommit_Java_ValidatesRunner_Spark.json]: @je-ik 
Thanks! I rollbacked trigger files!

**Comment by je-ik** [comment]
Thanks @twosom!

**Pull request merged**